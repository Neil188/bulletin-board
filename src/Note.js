import React, { Component } from 'react';
import Draggable from 'react-draggable';
import FaPencil from 'react-icons/lib/fa/pencil';
import FaTrash from 'react-icons/lib/fa/trash';
import FaFloppyO from 'react-icons/lib/fa/floppy-o';

class Note extends Component {
    constructor(props) {
        super(props);
        this.state = {
            editing: false,
        };
        this.edit = this.edit.bind(this);
        this.randomBetween = this.randomBetween.bind(this);
        this.remove = this.remove.bind(this);
        this.save = this.save.bind(this);
        this.renderForm = this.renderForm.bind(this);
        this.renderDisplay = this.renderDisplay.bind(this);
    }

    componentWillMount() {
        this.style = {
            right: this.randomBetween(0, window.innerWidth - 150, 'px'),
            top: this.randomBetween(0, window.innerHeight - 150, 'px'),
            transform: `rotate( ${this.randomBetween(-25, 25, 'deg')} )`,
        };
    }

    shouldComponentUpdate( nextProps, nextState ) {
        return (
            this.props.children !== nextProps.children ||
            this.state !== nextState
        );
    }

    componentDidUpdate() {
        let textArea;
        if (this.state.editing) {
            textArea = this.newText;
            textArea.focus();
            textArea.select();
        }
    }

    edit() {
        this.setState({
            editing: true,
        });
    }

    randomBetween(x, y, s) {
        return x + Math.ceil( Math.random() * (y-x) ) + s;
    }

    remove() {
        this.props.onRemove(this.props.index);
    }

    save(e) {
        e.preventDefault();
        this.props.onChange(this.newText.value, this.props.index);
        this.setState({
            editing: false,
        });
    }

    renderForm() {
        return (
            <Draggable disabled='true'>
                <div>
                    <div className='note' style={this.style}>
                        <form onSubmit={this.save}>
                            <textarea
                                ref={ input => {this.newText= input;} }
                                defaultValue={this.props.children}
                                maxLength={40}
                                cols={18}
                                rows={3}
                            />
                            <button id='save'>
                                <FaFloppyO />
                            </button>
                        </form>
                    </div>
                </div>
            </Draggable>
        );
    }

    renderDisplay() {
        return (
            <Draggable>
                <div>
                    <div className='note' style={this.style}>
                        <p>{this.props.children}</p>
                        <span>
                            <button
                                onClick={this.edit}
                                id='edit'
                            >
                                <FaPencil />
                            </button>
                            <button
                                onClick={this.remove}
                                id='remove'
                            >
                                <FaTrash />
                            </button>
                        </span>
                    </div>
                </div>
            </Draggable>
        );
    }

    render() {
        return this.state.editing ?
            this.renderForm()
            : this.renderDisplay();
    }
}

export default Note;