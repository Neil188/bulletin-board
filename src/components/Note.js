import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Draggable from 'react-draggable';
import FaPencil from 'react-icons/lib/fa/pencil';
import FaTrash from 'react-icons/lib/fa/trash';
import FaFloppyO from 'react-icons/lib/fa/floppy-o';
import { calcRandomBetween } from '../utils/random';

class Note extends Component {

    state = {
        editing: false,
    };

    componentWillMount() {
        this.style = {
            right: calcRandomBetween(0, window.innerWidth - 150, 'px'),
            top: calcRandomBetween(0, window.innerHeight - 150, 'px'),
            transform: `rotate( ${calcRandomBetween(-25, 25, 'deg')} )`,
        };
    }

    shouldComponentUpdate( nextProps, nextState ) {
        return (
            this.props.children !== nextProps.children ||
            this.state !== nextState
        );
    }

    // componentDidUpdate() {
    //     let textArea;
    //     if (this.state.editing) {
    //         textArea = this.newText;
    //         // textArea.focus();
    //         // textArea.select();
    //     }
    // }

    edit = () => {
        this.setState(() => ({
            editing: true,
        }));
    }

    remove = () =>
        this.props.onRemove(this.props.index);

    save = (e) => {
        e.preventDefault();
        this.props.onChange(this.newText.value, this.props.index);
        this.setState(() => ({
            editing: false,
        }));
    }

    renderForm = () => (
        <Draggable disabled>
            <div>
                <div className='note' style={this.style}>
                    <form onSubmit={this.save}>
                        <textarea
                            ref={ input => {this.newText= input;} }
                            defaultValue={this.props.children}
                            maxLength={40}
                            rows={4}
                        />
                        <button
                            className='save'
                            aria-label='save'
                        >
                            <FaFloppyO />
                        </button>
                    </form>
                </div>
            </div>
        </Draggable>
    )

    renderDisplay = () =>(
        <Draggable>
            <div>
                <div className='note' style={this.style}>
                    <p>{this.props.children}</p>
                    <span>
                        <button
                            onClick={this.edit}
                            className='edit'
                            aria-label='edit'
                        >
                            <FaPencil />
                        </button>
                        <button
                            onClick={this.remove}
                            className='remove'
                            aria-label='remove'
                        >
                            <FaTrash />
                        </button>
                    </span>
                </div>
            </div>
        </Draggable>
    )

    render() {
        return this.state.editing ?
            this.renderForm()
            : this.renderDisplay();
    }
}

Note.propTypes = {
    index: PropTypes.number.isRequired,
    children: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
};

export default Note;
