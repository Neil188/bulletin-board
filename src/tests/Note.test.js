import React from 'react';
import { create } from 'react-test-renderer';
import ReactDOM from 'react-dom';
import Note from '../components/Note';

jest.mock('../utils/random');

describe('Note snapshots', () => {
    test('Note snapshot test', () => {
        const component = create(
            <Note index={0} onChange={() => null} onRemove={() => null} test>
                Test
            </Note>,
            {
                createNodeMock: (element) => {
                    if (element.type === 'textarea') {
                    // mock focus/select functions
                        return {
                            select: () => null,
                            focus: () => null,
                        };
                    }
                    return null;
                },
            }
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // simulate edit click
        const testInstance = component.root;
        testInstance.findByProps({className:'edit'}).props.onClick();
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

    });
});

describe('Check for render failures', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <Note index={0} onChange={() => null} onRemove={() => null} test>
                Test
            </Note>
            , div);
    });
});