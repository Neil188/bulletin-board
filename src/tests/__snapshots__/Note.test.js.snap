// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simulate actions Click edit button 1`] = `
<Draggable
  allowAnyClick={false}
  axis="both"
  bounds={false}
  cancel={null}
  defaultClassName="react-draggable"
  defaultClassNameDragged="react-draggable-dragged"
  defaultClassNameDragging="react-draggable-dragging"
  defaultPosition={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  disabled={true}
  enableUserSelectHack={true}
  grid={null}
  handle={null}
  offsetParent={null}
  onDrag={[Function]}
  onMouseDown={[Function]}
  onStart={[Function]}
  onStop={[Function]}
  position={null}
  transform={null}
>
  <div>
    <div
      className="note"
      style={
        Object {
          "right": 0,
          "top": 0,
          "transform": "rotate( 0 )",
        }
      }
    >
      <form
        onSubmit={[Function]}
      >
        <textarea
          defaultValue="Test"
          maxLength={40}
          rows={4}
        />
        <button
          aria-label="save"
          className="save"
        >
          <FaFloppyO />
        </button>
      </form>
    </div>
  </div>
</Draggable>
`;

exports[`Snapshots Note snapshot test 1`] = `
<Draggable
  allowAnyClick={false}
  axis="both"
  bounds={false}
  cancel={null}
  defaultClassName="react-draggable"
  defaultClassNameDragged="react-draggable-dragged"
  defaultClassNameDragging="react-draggable-dragging"
  defaultPosition={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  disabled={false}
  enableUserSelectHack={true}
  grid={null}
  handle={null}
  offsetParent={null}
  onDrag={[Function]}
  onMouseDown={[Function]}
  onStart={[Function]}
  onStop={[Function]}
  position={null}
  transform={null}
>
  <div>
    <div
      className="note"
      style={
        Object {
          "right": 0,
          "top": 0,
          "transform": "rotate( 0 )",
        }
      }
    >
      <p>
        Test
      </p>
      <span>
        <button
          aria-label="edit"
          className="edit"
          onClick={[Function]}
        >
          <FaPencil />
        </button>
        <button
          aria-label="remove"
          className="remove"
          onClick={[Function]}
        >
          <FaTrash />
        </button>
      </span>
    </div>
  </div>
</Draggable>
`;
